crates/ironrdp-pdu/src/geometry.rs

```diff
@@ -1,161 +1,210 @@
 use std::cmp::{max, min};
-use std::io;
 
-use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use crate::cursor::ReadCursor;
+use crate::{PduDecode, PduEncode, PduResult};
 
-pub trait RectangleImpl {
-    const WIDTH_VALIDATION_ERROR: &'static str;
-    const HEIGHT_VALIDATION_ERROR: &'static str;
+pub trait Rectangle: Sized {
+    fn width(&self) -> u16;
+    fn height(&self) -> u16;
+    fn from_base(rect: BaseRectangle) -> Self;
+    fn to_base(&self) -> BaseRectangle;
 
-    fn calculate_width(left: u16, right: u16) -> u16;
-    fn calculate_height(top: u16, bottom: u16) -> u16;
+    fn empty() -> Self {
+        Self::from_base(BaseRectangle::empty())
+    }
+
+    fn union_all(rectangles: &[Self]) -> Self {
+        let result = rectangles.iter().fold(BaseRectangle::empty(), |a, b| {
+            let b = b.to_base();
+            BaseRectangle {
+                left: min(a.left, b.left),
+                top: min(a.top, b.top),
+                right: max(a.right, b.right),
+                bottom: max(a.bottom, b.bottom),
+            }
+        });
+
+        Self::from_base(result)
+    }
+
+    fn intersect(&self, other: &Self) -> Option<Self> {
+        let a = self.to_base();
+        let b = other.to_base();
+
+        let result = BaseRectangle {
+            left: max(a.left, b.left),
+            top: max(a.top, b.top),
+            right: min(a.right, b.right),
+            bottom: min(a.bottom, b.bottom),
+        };
+
+        if result.left < result.right && result.top < result.bottom {
+            Some(Self::from_base(result))
+        } else {
+            None
+        }
+    }
 
-    fn is_valid_width(left: u16, right: u16) -> bool;
-    fn is_valid_height(top: u16, bottom: u16) -> bool;
+    fn union(&self, other: &Self) -> Self {
+        let a = self.to_base();
+        let b = other.to_base();
+
+        let result = BaseRectangle {
+            left: min(a.left, b.left),
+            top: min(a.top, b.top),
+            right: max(a.right, b.right),
+            bottom: max(a.bottom, b.bottom),
+        };
+
+        Self::from_base(result)
+    }
 }
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct BaseRectangle<I: RectangleImpl> {
+pub struct InclusiveRectangle {
     pub left: u16,
     pub top: u16,
     pub right: u16,
     pub bottom: u16,
-    _phantom_data: std::marker::PhantomData<I>,
 }
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct InclusiveRectangleImpl;
+pub struct ExclusiveRectangle {
+    pub left: u16,
+    pub top: u16,
+    pub right: u16,
+    pub bottom: u16,
+}
 
-impl RectangleImpl for InclusiveRectangleImpl {
-    const WIDTH_VALIDATION_ERROR: &'static str = "width cannot be negative (inclusive rectangle)";
-    const HEIGHT_VALIDATION_ERROR: &'static str = "height cannot be negative (inclusive rectangle)";
+pub struct BaseRectangle {
+    left: u16,
+    top: u16,
+    right: u16,
+    bottom: u16,
+}
```

Comment:
I understand this struct exists so blanket implementations can work with this concrete struct instead of the generic `Self`. Is that correct?
I think it’s smart. I see two potential alternatives:
- replace `to_base` and `from_base` by a `fn new(left: u16, top: u16, width: u16, height: u16)`, returning an `Option<Self>` or `Result<Self>` (e.g.: inclusive rectangles can’t have a width or height of 0, because it implies `bottom - top + 1 = 0 ⇔ bottom = top - 1 ⇒ top < bottom` which is an impossible rectangle). It’s generally possible to write a blanket function based solely on `left`, `top`, `width` and `height` (we can get an exclusive `right` bound by computing `left + width`.  From there, we can build a `Self` again using the `new` method from the `Rectangle` trait.
- seal and hide `to_base` and `from_base` along the `BaseRectangle`, ideally it should not appear to be part of the the public API as it is an implementation detail (we are mostly there). Maybe you can have a look at this and see if you come up with something better: https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust

To be honest, I don’t know what is the best here, so feel free to choose or to ignore. What do you think?

Reply:
Oh! I didn't know that public traits could have trait bound which is not publicly exported! I like this idea, I agree that `to_base` and `from_base` in the public interface are ugly

Reply:
moved to `to_base` and `from_base` to sealed trait